src/Ocelot/ServiceDiscovery/Providers/PolingConsulServiceDiscoveryProvider.cs

```diff
@@ -15,42 +15,109 @@ namespace Ocelot.ServiceDiscovery.Providers
 {
     public class PollingConsulServiceDiscoveryProvider : IServiceDiscoveryProvider
     {
+        private readonly ConsulRegistryConfiguration _config;
         private readonly IOcelotLogger _logger;
-        private readonly IServiceDiscoveryProvider _consulServiceDiscoveryProvider;
-        private readonly Timer _timer;
-        private bool _polling;
+        private readonly IConsulClient _consul;
+        private const string VersionPrefix = "version-";
         private List<Service> _services;
-        private string _keyOfServiceInConsul;
 
-        public PollingConsulServiceDiscoveryProvider(int pollingInterval, string keyOfServiceInConsul, IOcelotLoggerFactory factory, IServiceDiscoveryProvider consulServiceDiscoveryProvider)
-        {;
+        private int _pollingInterval;
+        private ulong _waitIndex;
+
+        public PollingConsulServiceDiscoveryProvider(int pollingInterval, ConsulRegistryConfiguration config, IOcelotLoggerFactory factory, IConsulClientFactory clientFactory)
+        {
+            ;
+            _pollingInterval = pollingInterval == 0 ? 10000 : pollingInterval;
             _logger = factory.CreateLogger<PollingConsulServiceDiscoveryProvider>();
-            _keyOfServiceInConsul = keyOfServiceInConsul;
-            _consulServiceDiscoveryProvider = consulServiceDiscoveryProvider;
-            _services = new List<Service>();
-        
-            _timer = new Timer(async x =>
+
+            _config = config;
+            _consul = clientFactory.Get(_config);
+
+            Task.Factory.StartNew(async () =>
             {
-                    if(_polling)
-                    {
-                        return;
-                    }
-                    
-                    _polling = true;
-                    await Poll();
-                    _polling = false;
-                
-            }, null, pollingInterval, pollingInterval);
+                await Poll();
+            });
         }
 
-        public Task<List<Service>> Get()
-        {   
-            return Task.FromResult(_services);
+        public async Task<List<Service>> Get()
+        {
+            if (_services == null)
+            {
+                _services = await GetService();
+            }
+
+            return _services;
+        }
+
+        public async Task<List<Service>> GetService()
+        {
+            QueryResult<ServiceEntry[]> queryResult = null;
+            List<Service> services = new List<Service>();
+
+            if (_waitIndex == 0 || _pollingInterval == 0)
+            {
+                queryResult = await _consul.Health.Service(_config.KeyOfServiceInConsul, string.Empty, true);
+            }
+            else
+            {
+                // block queries
+                queryResult = await _consul.Health.Service(_config.KeyOfServiceInConsul, string.Empty, true,
+                    new QueryOptions() { WaitIndex = _waitIndex, WaitTime = TimeSpan.FromMilliseconds(_pollingInterval) });
```

Comment:
Does this waittime block the recursive call to Poll() in Poll()? I will pull the code and test this but I though I would ask to save some time

Reply:
yes, waittime block the recursive call to Poll() in Poll().